proposals/0046-optimistic-cluster-restart-automation.md

```diff
@@ -0,0 +1,321 @@
+optimistically confirmed.
+
+We need to prevent false negative at all costs, because we can't rollback an 
+optimistically confirmed block. However, false positive is okay. Because when 
+we select the heaviest fork in the next step, we should see all the potential 
+candidates for optimistically confirmed slots, there we can count the votes and
+remove some false positive cases.
+
+However, it's also overkill to repair every block presented by others. When
+`LastVotedForkSlots` messages are being received and aggregated, a validator
+can categorize blocks missing locally into 3 categories: ignored, must-have, 
+and unsure. Depending on the stakes of validators currently in restart, some
+slots with too few stake can be safely ignored, some have enough stake they
+should definitely be repaired, and the rest would be undecided pending more
+confirmations.
+
+Assume `RESTART_STAKE_THRESHOLD` is 80% and that 5% restarted validators can
+change their votes from what they voted before the restart due to mistakes or 
+malicious behavior.
+
+When only 5% validators are in restart, everything is in "unsure" category.
+
+When 67% validators are in restart, any slot with less than
+67% - 5% - (100-67%) = 29% is in "ignored" category, because even if all 
+validators join the restart, the slot will not get 67% stake. When this 
+threshold is less than 33%, we temporarily put all blocks with >33% stake into  
+"must-have" category to speed up repairing. Any slot with between 29% and 33% 
+stake is "unsure".
+
+When 80% validators are in restart, any slot with less than
+67% - 5% - (100-80%) = 42% is in "ignored" category, the rest is "must-have".
+
+From above examples, we can see the "must-have" threshold changes dynamically 
+depending on how many validators are in restart. The main benefit is that a
+block will only move from "must-have/unsure" to "ignored" as more validators 
+join the restart, not vice versa. So the list of blocks a validator needs to
+repair will never grow bigger when more validators join the restart.
+
+Once the validator gets LastVotedForkSlots, it can draw a line which are the
+"must-have" blocks. When all the "must-have" blocks are repaired and replayed,
+it can proceed to step 3.
+
+### 3. "silent repair phase": Gossip current heaviest fork
+
+The main goal of this step is to "vote" the heaviest fork to restart from.
+
+We use a new Gossip message `HeaviestFork`, its fields are:
+
+* `slot`: `u64` slot of the picked block.
+* `hash`: `Hash` bank hash of the picked block.
+* `received`: `u8` total percentage of stakes of the validators it received
+`HeaviestFork` messages from.
+
+After receiving `LastVotedForkSlots` from the validators holding stake more 
+than  `RESTART_STAKE_THRESHOLD` and repairing slots in "must-have" category,
+replay all blocks and pick the heaviest fork as follows:
+
+1. For all blocks with more than 67% votes, they must be on picked fork.
```

Comment:
1. "For all blocks with more than 67% votes", is this 67% of the `LastVotedForkSlots` or just votes we see in blocks?
2. What is "picked fork" here? Is this the heaviest fork?